### ISSUE TITLE
[XMaterial] - Missing texture block for beds in pre-1.12

### ISSUE THREAD

**Issue by reussy** [CryptoMorin/XSeries#144]
**Description**
I have a menu where I have placed an item with the material RED_BED, the problem is that in 1.8 client it shows a strange texture and in 1.16 client it shows correctly.
![Screenshot_24](https://user-images.githubusercontent.com/63428864/141658706-556b44a7-10c4-4451-82b9-040585cfd97f.png)
![Screenshot_25](https://user-images.githubusercontent.com/63428864/141658708-9dc0a4df-cf09-481c-a782-c882e1627753.png)

`public ItemStack createItem(boolean setGlow, XMaterial material, int amount, String name, List<String> lore, String value) throws IllegalArgumentException {

        ItemStack item = material.parseItem();
        assert item != null;
        item.setAmount(amount);
        ItemMeta meta = item.getItemMeta();
        assert meta != null;
        meta.setDisplayName(ChatColor.translateAlternateColorCodes('&', name));
        meta.setLore(lore);

        if (setGlow) {
            item.addUnsafeEnchantment(Enchantment.DAMAGE_ALL, 10);
            meta.addItemFlags(ItemFlag.HIDE_ENCHANTS);
            item.setItemMeta(meta);
        }
        item.setItemMeta(meta);

        if (item.getType() == XMaterial.PLAYER_HEAD.parseMaterial()) {

            SkullMeta texture = (SkullMeta) item.getItemMeta();
            SkullUtils.applySkin(texture, value);
            item.setItemMeta(texture);
        }

        return item;
    }`

I did a try catch for any problems and it is triggered only when I use "BED" instead of "RED_BED".
![imagen_2021-11-13_145315](https://user-images.githubusercontent.com/63428864/141658772-0d9011c3-ed41-4b56-ae40-095bf507f1f8.png)

**Version**
Server software: PaperSpigot
Server version 1.8.8
![imagen_2021-11-13_145410](https://user-images.githubusercontent.com/63428864/141658792-f6c23e44-2cc0-4776-8973-42b6f3d46000.png)

**Comment by CryptoMorin** [comment]
I just want to clarify this issue. You pretty much showed every part of your code except that one exact important part on how you obtained the `material` variable, but I figured it out.

Now I want to just leave a comprehensible explanation here since this is a complicated issue.

Colored Minecraft beds were added in **1.12** and they're pretty much the only item that had its *data value* changed, since there were only one bed before **1.12** (thus the data value 0) and it was changed to **14** as a result of the *default* bed item being red.

```java
Player player = event.getPlayer();
ItemStack a = XMaterial.RED_BED.parseItem();
ItemMeta aa = a.getItemMeta(); aa.setDisplayName("A"); a.setItemMeta(aa);

ItemStack b = XMaterial.matchXMaterial("BED").get().parseItem();
ItemMeta bb = b.getItemMeta(); bb.setDisplayName("B"); b.setItemMeta(bb);

ItemStack c = XMaterial.matchXMaterial("RED_BED").get().parseItem();
ItemMeta cc = c.getItemMeta(); cc.setDisplayName("C"); c.setItemMeta(cc);

ItemStack d = XMaterial.GREEN_BED.parseItem();
ItemMeta dd = d.getItemMeta(); dd.setDisplayName("D"); d.setItemMeta(dd);

player.getInventory().addItem(a, b, c, d);
```

There are several issues with this test:
**1.** The code above gives the *missing texture block* for all cases when used before **1.12**, but it can still be placed and used like a normal bed.
**2.** You will however get a black bed for case **B** in **1.12+** as a result of XMaterial's alphabetic enum ordering search.
**3.** Using any of the other colored beds for versions before **1.12** gives you the same *missing texture block* issue.

This is actually rather easy to fix because of the separated `Data` class we're able to manipulate the *data value* for red beds like so:
```java
RED_BED(supports(12) ? 14 : 0, "BED_BLOCK", "BED"),
``` 
However, this only fixes the issue for case **A** and **C**, but you can do the same to all the other materials.
```java
// Other than matchXMaterial("BLACK_BED") and XMaterial.BLACK_BED.parseItem() also fixes matchXMaterial("BED")
BLACK_BED(supports(12) ? 15 : 0, "BED_BLOCK", "BED"),
GREEN_BED(supports(12) ? 13 : 0, "BED_BLOCK", "BED"),
```
This fixes all cases for **pre-1.12**, but case **B** will give you **BLACK_BED** in **1.12** which is kinda undesirable, but it's not important.

Another solution would be the following which fixes the last issue as well.
```java
static {
    if (!supports(12)) {
        NAME_CACHE.put("BED", XMaterial.RED_BED);
        for (XMaterial mat : new XMaterial[]{BLACK_BED, WHITE_BED, GREEN_BED, ...}) NAME_CACHE.put(mat.name(), XMaterial.RED_BED);
    }
}
```
But this requires `NAME_CACHE` to be a non-expiring cache which can (rarely) cause unnecessary memory allocations.

I think the first solution seems the most reasonable in terms of performance, server owners should use "RED_BED" if their server is running pre-1.12 versions.

**Comment by CryptoMorin** [comment]
Followed the first solution in v3.5.0
